File: authme-android/app/src/main/java/org/authme/android/service/AuthMeSign.java
/*
 * Copyright 2011 Berin Lautenbach
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package org.authme.android.service;

import android.util.Base64;
import android.util.Log;

import org.authme.event.DeviceAddedEvent;
import org.authme.event.ResponseEvent;
import org.authme.event.SignatureSeedInfoEvent;
import org.json.JSONException;
import org.json.JSONObject;

import java.security.KeyPair;
import java.security.SecureRandom;
import java.security.Signature;
import java.text.SimpleDateFormat;
import java.util.Date;

import timber.log.Timber;

/**
 * Created by IntelliJ IDEA.
 * User: berin
 * Date: 7/08/11
 * Time: 5:36 PM
 *
 * Signs a string using the master password.  RUNS ON CURRENT THREAD - always call on an already
 * backgrounded thread
 */
public class AuthMeSign implements AuthMeService.Callbacks {

    // For logging
    public static final String TAG = "AuthMeSign";

    private String toSign;
    private KeyPair keyPair;
    private SigningCallbacks callback = null;

    public String sigId;
    public String dateTime;
    public String signature;

    public AuthMeSign() {

    }

    public boolean doSign(String toSign, KeyPair keyPair, SigningCallbacks callback) {

        this.toSign = toSign;
        this.keyPair = keyPair;
        this.callback = callback;

        AuthMeService authme = new AuthMeService(MasterPassword.getInstance().getApplicationContext());
        if (!authme.getSignatureSeed(this))
            return false;

        /* Now we wait for the service to get back to us */
        return true;

    }

    private void finishSignature (SignatureSeedInfoEvent seedInfo) {

        sigId = seedInfo.getSigId();
        dateTime = seedInfo.getDateTime();

        Log.v(TAG, "Sig: " + sigId + " / DateTime: " + dateTime);
        Log.v(TAG, "Signing: " + toSign);

        String signData = sigId + dateTime + toSign;
        Log.v(TAG, "Full sign = (" + signData + ")");

        try {
            Signature sig = Signature.getInstance("SHA1withRSA");
            sig.initSign(keyPair.getPrivate(), new SecureRandom());

            sig.update(signData.getBytes());

            byte sigBytes[] = sig.sign();
            signature = Base64.encodeToString(sigBytes, Base64.NO_WRAP);

            if (callback != null)
                callback.onSignatureReturn(this);
        }

        catch(Exception ex) {
            Log.w(TAG, "Error signing data: " +ex.getMessage());
        }



    }

    /* My callback */

    public interface SigningCallbacks {

        void onSignatureReturn(AuthMeSign signature);

    }

    /* AuthMe service return */
    public void onAuthMeServiceReturn(ResponseEvent responseEvent) {

        Timber.d("Received response event in signature object");

        if (responseEvent == null || !responseEvent.getSuccess()) {
            Timber.d("Error in AuthMe response");
            return;
        };

        if (responseEvent instanceof SignatureSeedInfoEvent) {
            finishSignature((SignatureSeedInfoEvent) responseEvent);
        }

    }
}

File: authme-iphone/AuthMe/Src/Service/AuthMeSign.swift
/*
 *
 * Copyright 2015 Berin Lautenbach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

//
//  AuthMeSign.swift
//  AuthMe
//
//  Created by Berin Lautenbach on 30/01/2016.
//  Copyright © 2016 Berin Lautenbach. All rights reserved.
//

import Foundation

class AuthMeSign : AuthMeServiceDelegate {
    
    var logger = Log()
    
    var sigId = ""
    var dateTime = ""
    var signature = ""
    var toSign = ""
    var delegate: AuthMeSignDelegate? = nil
    var opaqueData: AnyObject? = nil
    
    var keyPair : RSAKey? = nil
    
    var masterPassword: MasterPassword
    var authme: AuthMeService
    
    init() {
        masterPassword = MasterPassword.getInstance()
        authme = AuthMeService()
    }

    func doSign(_ toSign: String, keyPair: RSAKey, delegate: AuthMeSignDelegate, withOpaqueData opaqueData: AnyObject?) {
        self.toSign = toSign
        self.keyPair = keyPair
        self.delegate = delegate
        self.opaqueData = opaqueData
        
        if self.keyPair == nil {
            return
        }
        
        authme.getSignatureSeed(self, withOpaqueData: opaqueData)
        
    }
    
    
    func service(_ service: AuthMeService, didCompletOperation operation: AuthMeServiceOperation, withOpaqueData opaqueData: AnyObject?) {
        
        if operation.operationType != AuthMeService.AuthMeOperationType.getSignatureSeed {
            logger.log(.error, message: "Somehow got a authme return that isn't a signature seed into a signer object")
                return
        }
        
        var success = false
        
        if operation.statusCode == 200 {
            if let readData = operation.returnData {
                let json = (try! JSONSerialization.jsonObject(with: readData as Data, options: JSONSerialization.ReadingOptions.mutableContainers)) as! NSDictionary
                if (json.object(forKey: "sigId") as? NSString != nil) && (json.object(forKey: "dateTime") as? NSString != nil) {
                    sigId = json.value(forKey: "sigId") as! String
                    dateTime = json.value(forKey: "dateTime") as! String
                    
                    if keyPair != nil {
                        //signature = keyPair!.sign("\(sigId)\(toSign)".dataUsingEncoding(NSUTF8StringEncoding))
                        signature = keyPair!.sign("\(sigId)\(dateTime)\(toSign)".data(using: String.Encoding.utf8))
                        success = true
                    }
                }
            }
        }
        
        self.delegate?.signerDidComplete(self, didSucceed: success, withOpaqueData: opaqueData)
    }
    
}
